core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactorySpec.js

```diff
@@ -0,0 +1,133 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview unit tests for number with units object type factory service.
+ */
+describe('NumberWithUnitsObjectFactory', function() {
+  beforeEach(module('oppia'));
+
+  describe('number with units object factory', function() {
+    var NumberWithUnits = null;
+    var Units = null;
+    var Fraction = null;
+    var errors = null;
+
+    beforeEach(inject(function($injector) {
+      NumberWithUnits = $injector.get('NumberWithUnitsObjectFactory');
+      Units = $injector.get('UnitsObjectFactory');
+      Fraction = $injector.get('FractionObjectFactory');
+      errors = $injector.get('NUMBER_WITH_UNITS_PARSING_ERRORS');
+    }));
+
+    it('should convert units to dict format', function() {
+      expect(new Units('kg / kg^2 K mol / (N m s^2) K s').toDict()).toEqual(
+        {kg: -1, K: 2, mol: 1, N: -1, m: -1, s: -1});
+      expect(new Units('mol/(kg / (N m / s^2)').toDict()).toEqual(
+        {mol: 1, kg: -1, N: 1, m: 1, s: -2});
+      expect(new Units('kg per kg^2 K mol per (N m s^2) K s').toDict()).toEqual(
+        {kg: -1, K: 2, mol: 1, N: -1, m: -1, s: -1});
+    });
+
+    it('should convert units from dict to string format', function() {
+      expect(Units.fromDictToString(
+        {kg: -1, K: 2, mol: 1, N: -1, m: -1, s: -1})).toBe(
+        'kg^-1 K^2 mol^1 N^-1 m^-1 s^-1');
+      expect(Units.fromDictToString(
+        {mol: 1, kg: -1, N: 1, m: 1, s: -2, K: -1})).toBe(
+        'mol^1 kg^-1 N^1 m^1 s^-2 K^-1');
+    });
+
+    it('should convert units from string to lexical format', function() {
+      expect(Units.stringToLexical('kg per kg^2 K mol / (N m s^2) K s'
+      )).toEqual(['kg', '/', 'kg^2', '*', 'K', '*', 'mol', '/', '(', 'N', '*',
+                  'm', '*', 's^2', ')', 'K', '*', 's']);
+      expect(Units.stringToLexical('kg (K mol) m/s^2 r t / (l/ n) / o'
+      )).toEqual(['kg', '(', 'K', '*', 'mol', ')', 'm', '/', 's^2', '*', 'r',
+                  '*', 't', '/', '(', 'l', '/', 'n', ')', '/', 'o']);
+      expect(Units.stringToLexical('mol per (kg / (N m / s^2)*K)'
+      )).toEqual(['mol', '/', '(', 'kg', '/', '(', 'N', '*', 'm', '/', 's^2',
+                  ')', '*', 'K', ')']);
+    });
+
+    it('should convert itself to a string', function() {
+      expect(new NumberWithUnits('real', 2.02, '', 'm / s^2').toString()).toBe(
+        '2.02 m / s^2');
+      expect(new NumberWithUnits('real', 2.02, '', 'Rs').toString()).toBe(
+        'Rs 2.02');
+      expect(new NumberWithUnits('real', 2, '', '').toString()).toBe('2');
+      expect(new NumberWithUnits('fraction', '', new Fraction(
+        true, 0, 4, 3), 'm / s^2').toString()).toBe('-4/3 m / s^2');
```

Comment:
the unit should be a UnitsObjectFactory object and not a string.

Reply:
I'll modify it. Btw units object is itself storing units as string for your confirmation.

Reply:
Then you have it incorrect as the unit should be stored in dict format. That is how it is stored in the backend. So fix that. Basically, when you run toDict method on this object, you should get an object which imitates its backend storage schema.

Reply:
Done, I've changed it to Units object. Yes, using `toDict` method, we gets an object which imitates its backend storage schema.